### PR TITLE
fix(liquidator): de-spam failure notifications + buffer eligibility cap

### PR DESCRIPTION
**Pull request by vitaliybezz**
## What

Two coupled fixes for the recurring `❌ Liquidation Failed` Telegram spam (e.g. the `ixlm-ixlmusdc` notification firing every 10 minutes for the same borrower with the same on-chain `Attempt to liquidate more collateral than is currently eligible` panic).

## Fixes

**1. Notification dedup (stops the spam)**
- Same `(market, borrower, error class)` is suppressed within a configurable cooldown — default **24h** (`FAILURE_NOTIFICATION_COOLDOWN_HOURS`).
- A successful liquidation **immediately clears** the cooldown for that borrower, so the next genuine failure (of any kind) still alerts.
- `TransactionFailed` is sub-classified by the contract panic substring (`excessive_liquidation`, `offer_too_low`, `not_eligible`, `value_calc_failure`, `tx_timeout`, `tx_failed_other`) so distinct failures still get their own notification.

**2. Cap-aware safety buffer (prevents most of the reverts in the first place)**
- New `LIQUIDATABLE_CAP_BUFFER_BPS = 300` (3%) applied to `liquidatable_collateral` when it's the binding constraint in both `PercentageLiquidationStrategy` and `FixedAmountLiquidationStrategy`.
- Rationale: the contract's eligibility formula amplifies small collateral-price drift between scan and tx submission. The observed mismatch (`37,818,981 requested > 34,516,659 available` ≈ 9.6%) typically traces back to a ~1.6% collateral price move during the 10-minute scan cycle. 3% covers the common case while leaving most of the eligible collateral liquidatable.

## Tunables

| Env var | Default |
|---|---|
| `FAILURE_NOTIFICATION_COOLDOWN_HOURS` | `24` |

## Test plan

- [x] `cargo nextest run -p templar-liquidator` (110 passed, includes 5 dedup + 5 classifier + 3 buffer tests)
- [x] `cargo clippy -p templar-liquidator --no-deps -- -D warnings`
- [x] `cargo fmt --check`
- [ ] Deploy + monitor Telegram channel for 24h — verify the recurring `ixlm-ixlmusdc` spam goes silent and (separately) that real new failures still fire

- - -
This change is [Reviewable](https://reviewable.io/reviews/Templar-Protocol/contracts/449)

### CODE REVIEW

**Review by royalf00l — APPROVED** [review]

File: service/liquidator/src/config.rs [hunk 1]
```diff
@@ -359,7 +369,9 @@ impl Args {
                 panic!("TELEGRAM_BOT_TOKEN and TELEGRAM_CHAT_ID must both be set or both be empty");
             }
         };
-        let notifier = Arc::new(Notifier::new(telegram_config));
+        let failure_cooldown =
+            std::time::Duration::from_secs(self.failure_notification_cooldown_hours * 3600);
+        let notifier = Arc::new(Notifier::with_cooldown(telegram_config, failure_cooldown));
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/config.rs:374]: Skipping. Your own analysis script printed `Max safe FAILURE_NOTIFICATION_COOLDOWN_HOURS without overflow: 5_124_095_576_030_431` — that's ~5.8 × 10¹¹ years. Operators won't accidentally set a value within ~10 orders of magnitude of that. Adding `checked_mul + expect` here is theatre without a real failure mode.

File: service/liquidator/src/liquidator.rs [hunk 2]
```diff
@@ -175,6 +175,53 @@ impl LiquidatorError {
...
+    } else if msg.contains("Liquidation offer too low") {
+        "offer_too_low"
+    } else if msg.contains("not eligible for liquidation") {
+        "not_eligible"
+    } else if msg.contains("Failed to calculate value of collateral") {
+        "value_calc_failure"
+    } else if msg.contains("Timeout") || msg.contains("timeout") {
+        "tx_timeout"
+    } else {
+        "tx_failed_other"
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs]: Skipping. The matched strings come from `thiserror` derives on `LiquidationError` in `common/src/borrow.rs` with fixed casing — there's no drift to defend against. The `Timeout | timeout` redundancy is a fallback in case some other layer wraps the message; adding a per-call `to_ascii_lowercase()` allocation for stable strings is unnecessary work.

File: service/liquidator/src/notifier.rs [hunk 3]
```diff
@@ -52,6 +54,15 @@ pub type SharedNotifier = Arc<Notifier>;
...
 
+/// Default cooldown for repeated identical failure notifications.
+///
+/// Same (market, borrower, `error_kind`) within this window is suppressed.
+pub const DEFAULT_FAILURE_NOTIFY_COOLDOWN: Duration = Duration::from_secs(24 * 60 * 60);
+
+/// Dedup key for failure notifications. Stable across rounds for the same
+/// (market, borrower, error class).
+type DedupKey = (String, String, String);
+
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs:70]: Skipping for now. The `&'static str` approach mirrors the existing `ErrorPhase::as_str()` pattern in the same file. Promoting to a typed enum across `Notifier`, `DedupKey`, and `LiquidatorError` would be churn without solving an actual bug we have. Happy to revisit if we accumulate more kinds.

File: service/liquidator/src/notifier.rs [hunk 4]
```diff
@@ -230,10 +253,59 @@ impl Notifier {
...
+        if !self.should_send_failure(market, borrower, error_kind) {
+            tracing::debug!(
+                market,
+                borrower,
+                error_kind,
+                "Liquidation failure notification suppressed by dedup"
+            );
+            return;
+        }
         self.spawn_send(format_liquidation_failed_message(market, borrower, error));
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs]: Fixed in c0b9dc59. `spawn_send` now returns a bool, and `notify_liquidation_failed` rolls back the dedup entry via `rollback_failure_dedup` if the message couldn't be queued. The disabled-Telegram case still records (returns `true`) since "no Telegram configured" is a no-op, not a drop. Telegram-side delivery failures are still swallowed by `send()` — that's a separate retry concern.

File: service/liquidator/src/liquidator.rs [hunk 5]
```diff
@@ -175,6 +175,53 @@ impl LiquidatorError {
...
+    /// Returns a stable, low-cardinality kind string used to dedupe
+    /// repeat failure notifications for the same `(market, borrower)`.
+    ///
+    /// `TransactionFailed` is further classified by the contract panic
+    /// substring so a "wrong amount" failure and an "offer too low" failure
+    /// each fire their own notification once.
+    #[must_use]
+    pub fn notification_kind(&self) -> &'static str {
+        match self {
+            Self::TransactionFailed(msg) => classify_transaction_failure(msg),
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs:256]: Fixed in c0b9dc59. `notification_kind()` now matches `LiquidationTransactionError(RpcError::TimeoutError(_, _)) => "tx_timeout"` before falling through to the generic `tx_submission_error`. Added two tests (`test_notification_kind_tx_submission_timeout`, `test_notification_kind_tx_submission_non_timeout`).

File: service/liquidator/src/config.rs [hunk 6]
```diff
@@ -359,7 +369,9 @@ impl Args {
                 panic!("TELEGRAM_BOT_TOKEN and TELEGRAM_CHAT_ID must both be set or both be empty");
             }
         };
-        let notifier = Arc::new(Notifier::new(telegram_config));
+        let failure_cooldown =
+            std::time::Duration::from_secs(self.failure_notification_cooldown_hours * 3600);
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/config.rs:373]: Skipping — same reasoning as the other overflow comment above. CodeRabbit's analysis itself computed the safe max as ~5.1 × 10¹⁵ hours, which is ~5.8 × 10¹¹ years. No realistic misconfiguration can hit it.

File: service/liquidator/src/liquidator.rs [hunk 7]
```diff
@@ -892,8 +939,16 @@ impl Liquidator {
...
-                Ok(LiquidationOutcome::Liquidated) => liquidated += 1,
-                Ok(LiquidationOutcome::NotLiquidatable) => not_liquidatable += 1,
+                Ok(LiquidationOutcome::Liquidated) => {
+                    self.notifier
+                        .clear_failure_dedup_for(self.market.as_ref(), account.as_ref());
+                    liquidated += 1;
+                }
+                Ok(LiquidationOutcome::NotLiquidatable) => {
+                    self.notifier
+                        .clear_failure_dedup_for(self.market.as_ref(), account.as_ref());
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs:1033]: Fixed in c0b9dc59. Split `LiquidationOutcome::NotLiquidatable` into `Healthy` (chain confirms the position is OK) and `Skipped` (inventory or strategy refused, position may still be liquidatable). Dedup is now cleared only on `Liquidated` or `Healthy`; `Skipped` and `Unprofitable` leave the dedup state intact so a recurring execution failure can't be re-alerted by a transient inventory blip.

File: service/liquidator/src/liquidator.rs [hunk 8]
```diff
@@ -892,8 +939,16 @@ impl Liquidator {
                 .liquidate(account.clone(), position, oracle_response.clone())
                 .await
             {
-                Ok(LiquidationOutcome::Liquidated) => liquidated += 1,
-                Ok(LiquidationOutcome::NotLiquidatable) => not_liquidatable += 1,
+                Ok(LiquidationOutcome::Liquidated) => {
+                    self.notifier
+                        .clear_failure_dedup_for(self.market.as_ref(), account.as_ref());
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs:1028]: Acknowledged but punting. This requires `LOOP_LIQUIDATION=true` and the rare case where an early iteration succeeds and a later one fails with a previously-seen error class. Worth a separate change that clears the dedup inside the loop on each successful iteration, but it's orthogonal to the spam fix and I'd rather not enlarge this PR.

File: service/liquidator/src/liquidation_strategy.rs [hunk 9]
```diff
@@ -539,6 +550,30 @@ mod tests {
...
+    fn test_apply_liquidatable_cap_buffer_zero() {
+        assert_eq!(apply_liquidatable_cap_buffer(0), 0);
+    }
+
+    #[test]
+    fn test_apply_liquidatable_cap_buffer_handles_observed_drift() {
+        // Observed mismatch: 37_818_981 requested > 34_516_659 available (≈9.6% drift).
+        // With the buffer applied to the (then) liquidatable_collateral, our request
+        // would be 33_481_159, comfortably under the chain's recomputed eligibility
+        // even if drift exceeds the buffer's 3%.
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidation_strategy.rs]: Fully agree, fixed in c0b9dc59. Replaced the misleading test with two honest ones:
- `test_apply_liquidatable_cap_buffer_covers_drift_up_to_buffer_size` — asserts the buffer protects against drift ≤ 3%.
- `test_apply_liquidatable_cap_buffer_insufficient_for_large_drift` — explicitly demonstrates that the observed 9% drift still exceeds the buffer; notification dedup is the safety net for that case.

Also updated the constant's docstring to note the buffer only covers drift up to its own size.

File: service/liquidator/src/liquidation_strategy.rs [hunk 10]
```diff
@@ -256,7 +271,8 @@ impl LiquidationStrategy for PercentageLiquidationStrategy {
         let target_collateral = if market_version == Some((1, 0, 0)) {
             position.collateral_asset_deposit.into()
         } else {
-            std::cmp::min(collateral_amount, liquidatable_collateral.into())
+            let buffered_cap = apply_liquidatable_cap_buffer(liquidatable_collateral.into());
+            std::cmp::min(collateral_amount, buffered_cap)
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidation_strategy.rs]: Fixed in 105eba54. Extracted `min_with_cap_buffer(desired, liquidatable)` and use it in both `PercentageLiquidationStrategy` and `FixedAmountLiquidationStrategy` (DRY). Added three tests asserting the helper returns the desired amount when below the cap, clips to the buffered cap when above, and clips edge-case values just one unit above the buffered boundary. A regression that removes the buffer call in either strategy will fail these tests via the shared helper.

File: service/liquidator/src/liquidator.rs [hunk 11]
```diff
@@ -175,6 +181,54 @@ impl LiquidatorError {
...
+/// survives small wording changes and surrounding receipt-id boilerplate.
+fn classify_transaction_failure(msg: &str) -> &'static str {
+    if msg.contains("Attempt to liquidate more collateral") {
+        "excessive_liquidation"
+    } else if msg.contains("Liquidation offer too low") {
+        "offer_too_low"
+    } else if msg.contains("not eligible for liquidation") {
+        "not_eligible"
+    } else if msg.contains("Failed to calculate value of collateral") {
+        "value_calc_failure"
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs]: Fixed in 105eba54. Added `test_notification_kind_value_calc_failure` asserting the "Failed to calculate value of collateral" panic substring routes to `NotificationKind::ValueCalcFailure`.

File: service/liquidator/src/notifier.rs [hunk 12]
```diff
@@ -230,8 +253,76 @@ impl Notifier {
...
+    ///
+    /// The dedup entry is recorded only when the send is actually accepted
+    /// by the in-flight semaphore; if the message is dropped due to overload,
+    /// the entry is rolled back so the next call can retry.
+    pub fn notify_liquidation_failed(
+        self: &Arc<Self>,
+        market: &str,
+        borrower: &str,
+        error_kind: &str,
+        error: &str,
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/notifier.rs:273]: Fair point — you're right, and citing AGENTS.md was the deciding factor. Reversed my earlier skip. In 105eba54, introduced a `NotificationKind` enum (20 variants) with `as_str()` + `Display`. `LiquidatorError::notification_kind()` returns the enum directly, the `DedupKey` carries the enum, and `notify_liquidation_failed` takes `NotificationKind` as a typed parameter. A typo or mismatched string at a caller now fails to compile rather than silently creating a fragmented dedup bucket.

The `as_str()` representation is locked by a test so dedup state survives rolling restarts.

File: service/liquidator/src/liquidation_strategy.rs [hunk 13]
```diff
@@ -424,7 +448,7 @@ impl LiquidationStrategy for FixedAmountLiquidationStrategy {
             position.collateral_asset_deposit.into()
         } else {
             let safe_collateral = (max_collateral * (10_000 - SAFETY_BUFFER_BPS)) / 10_000;
-            std::cmp::min(safe_collateral, liquidatable_u128)
+            min_with_cap_buffer(safe_collateral, liquidatable_u128)
         };
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidation_strategy.rs:460]: Fixed in afa483e4. Both strategies now check `if target_collateral == 0` after `min_with_cap_buffer` and return `Ok(None)` with a warning log. Practically this only fires when `liquidatable_collateral < 34` raw units (the 3% buffer rounds anything below that to zero), but the resulting `Some((1, 0))` would have been a logical bug regardless. Added a test (`test_min_with_cap_buffer_returns_zero_for_dust_cap`) pinning the boundary.

File: service/liquidator/src/liquidator.rs [hunk 14]
```diff
@@ -418,7 +536,7 @@ impl Liquidator {
                 return Ok(if loop_iteration > 1 {
                     LiquidationOutcome::Liquidated
                 } else {
-                    LiquidationOutcome::NotLiquidatable
+                    LiquidationOutcome::Healthy
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/liquidator.rs]: Good catch, fixed in afa483e4. Replaced the `let-else` with an explicit match:
- `Some(BorrowStatus::Liquidation(reason))` → proceed
- `Some(BorrowStatus::MaintenanceRequired)` → return `Skipped` (unhealthy but not liquidatable; dedup state stays intact)
- `Some(BorrowStatus::Healthy) | None` → return `Healthy` / `Liquidated` (clears dedup)

File: service/liquidator/src/config.rs [hunk 15]
```diff
@@ -178,6 +178,16 @@ pub struct Args {
...
 
+    /// Cooldown in hours for repeated "Liquidation Failed" notifications with
+    /// the same (market, borrower, error class). Successful liquidations
+    /// reset the cooldown for that borrower immediately.
+    #[arg(
+        long,
+        env = "FAILURE_NOTIFICATION_COOLDOWN_HOURS",
+        default_value_t = 24
+    )]
+    pub failure_notification_cooldown_hours: u64,
```
[the remark this answers is not in the document]
Comment (vitaliybezz) [reply service/liquidator/src/config.rs:189]: Fixed in afa483e4. Added `DEFAULT_FAILURE_NOTIFY_COOLDOWN_HOURS = 24` as the single source of truth in `notifier.rs`. The `Duration` constant derives from it (`DEFAULT_FAILURE_NOTIFY_COOLDOWN_HOURS * 60 * 60`), and the clap `default_value_t` references it via `crate::notifier::DEFAULT_FAILURE_NOTIFY_COOLDOWN_HOURS`. Test args helper also references the constant so a single change propagates everywhere.
